src/kemal/route_handler.cr

```diff
@@ -28,12 +30,22 @@ module Kemal
       @routes.find radix_path(verb, path)
     end
 
+    def lookup_route(request)
+      lookup_route request.override_method.as(String), request.path
+    end
+
+    def route_defined?(request)
+      lookup_route(request).found?
+    end
+
     # Processes the route if it's a match. Otherwise renders 404.
     private def process_request(context)
-      raise Kemal::Exceptions::RouteNotFound.new(context) unless context.route_defined?
-      content = context.route.handler.call(context)
+      raise Kemal::Exceptions::RouteNotFound.new(context) unless route_defined?(context.request)
+      tree_result = lookup_route(context.request)
+      context.request.url_params = tree_result.params
```

Comment:
We actually need to assign `url_params` before coming to `RouteHandler` in `HTTP::Server::Context` like we did before. Otherwise the custom middlewares before `RouteHandler` won't be able to access url params.

Reply:
Okay. Init handler then? This will mean an additional route lookup again :/

Reply:
@straight-shoota `InitHandler` won't cut it. What if they override the handler chain and don't use `InitHandler` at all?

Reply:
That would require some messing around with Kemal's internals . You can't simply skip `setup_init_handler`,  you'd have to call `app.setup` individually, then remove the init handler from `app.handlers` before calling `app.run`. I'd say this should not be a supported mode of operation and we should assume that `InitHandler` is always the first handler to be called in a proper Kemal application.

Reply:
Currently in `master`, you can do `Kemal.config.handlers = [MyHandler.new]` and totally customize the handler chain.